internal/integratedservices/services/common.go

```diff
@@ -0,0 +1,35 @@
+// Copyright © 2020 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package services
+
+import (
+	"reflect"
+
+	"emperror.dev/errors"
+	"github.com/mitchellh/mapstructure"
+
+	"github.com/banzaicloud/pipeline/internal/integratedservices"
+)
+
+// BindIntegratedServiceSpec binds an incoming integrated service specific raw spec (json) into the appropriate struct
+func BindIntegratedServiceSpec(inputSpec integratedservices.IntegratedServiceSpec, boundSpec interface{}) error {
+	if err := mapstructure.Decode(inputSpec, &boundSpec); err != nil {
+		return integratedservices.InvalidIntegratedServiceSpecError{
+			IntegratedServiceName: reflect.TypeOf(boundSpec).Name(),
+			Problem:               errors.WrapIf(err, "failed to bind integrated service spec").Error(),
```

Comment:
If the only thing we need from the error is the message (which does not seem like a good idea at first), then `fmt.Sprintf` might be better. We annotate the error with stack trace here, just to drop it one call later.

Reply:
let's discuss it and i'll add it to a new PR